stylix/testbed.nix

```diff
@@ -0,0 +1,95 @@
+{ pkgs, inputs, lib, ... }:
+
+let
+  commonModule = { config, ... }: {
+    users.users.testbed = {
+      description = "Testbed";
+      hashedPassword = "";
+      isNormalUser = true;
+    };
+
+    # The state version can safely track the latest release because the disk
+    # image is ephermal.
+    system.stateVersion = config.system.nixos.release;
+    home-manager.users.testbed.home.stateVersion = config.system.nixos.release;
+
+    virtualisation.vmVariant.virtualisation = {
+      # This is a maximum limit; the VM should still work if the host has fewer cores.
+      cores = 4;
+      memorySize = 2048;
+    };
+  };
+
+  autoload = builtins.concatLists
+    (lib.mapAttrsToList
+      (name: _:
+        let testbed = {
+          inherit name;
+          module = "${../modules}/${name}/testbed.nix";
+        };
+        in
+          lib.optional (builtins.pathExists testbed.module) testbed
+      )
+      (builtins.readDir ../modules));
+
+  makeTestbed =
+    testbed: stylix:
+    let
+      name = "testbed-${testbed.name}-${stylix.polarity}";
+
+      system = lib.nixosSystem {
+        inherit (pkgs) system;
+        modules = [
+          inputs.self.nixosModules.stylix
+          inputs.home-manager.nixosModules.home-manager
+          commonModule
+          testbed.module
+          {
+            inherit stylix;
+            system.name = name;
+          }
+        ];
+      };
+
+      script = pkgs.writeShellApplication {
+        inherit name;
+        text = ''
+          directory="''${XDG_CACHE_HOME:-$HOME/.cache}/stylix-testbed"
+          mkdir --parents "$directory"
+
+          # The disk image is only initialised if the file doesn't already exist,
+          # so we need to create a temporary directory and give a path within it
+          # rather than creating a temporary file. 
+          directory="$(mktemp --directory "$directory/XXXXXXXXXX")"
+
+          NIX_DISK_IMAGE="$directory/nixos.qcow2"
+          export NIX_DISK_IMAGE
+
+          clean() {
+            if rm --recursive "$directory"; then
+              echo 'Virtualisation disk image removed.'
+            fi
+          }
+          trap clean EXIT
```

Comment:
Why are the images stored in the `${XDG_CACHE_HOME:-$HOME/.cache}/stylix-testbed` directory when the `cleanup` function ensures that the contained directories are properly cleaned up? This unnecesssarily clutters the user's cache directory. Could we not create the directory directly with `mktemp --directory`:

```suggestion
          cleanup() {
            if rm --recursive "$image"; then
              echo 'Virtualisation disk image removed.'
            fi
          }

          # Create a temporary directory for containing the disk image, instead
          # of creating a temporary 'NIX_DISK_IMAGE' file because the
          # 'NIX_DISK_IMAGE' file must not exist as it will otherwise not be
          # written to.
          directory="$(mktemp --directory)"
          trap cleanup EXIT

          export NIX_DISK_IMAGE="$directory/nixos.qcow2"
```

Reply:
The reason why this ended up this way is as follows:

1. Originally, the images were stored as `/run/user/UID/stylix/hostname.qcow2` and cleaned up next time the testbed was started.
2. I moved this to `~/.cache` to avoid holding storing potentially large images in memory. The images were still cleaned up on the next run, so they had to be somewhere where the naming wasn't random.
3. I made it clean up at the end of each run, rather than the start of the next run, and introduced randomised names.

So yes, we can now safely use `/tmp` rather than `~/.cache`.

Reply:
> So yes, we can now safely use `/tmp` rather than `~/.cache`.

Alright!

As explained in the source code comment, it is an implementation detail that we are deleting the temporary directory. Based on that, should we replace the `echo` command with `verboseEcho`, similar to #316, or remove the `echo` command altogether?

If we use `verboseEcho`, then we should do something like:

```suggestion
          cleanup() {
            if string="$(rm --recursive --verbose "$directory")"; then
              verboseEcho "$string"
            fi
          }

          # Create a temporary directory for containing the disk image, instead
          # of creating a temporary 'NIX_DISK_IMAGE' file because the
          # 'NIX_DISK_IMAGE' file must not exist as it will otherwise not be
          # written to.
          directory="$(mktemp --directory)"
          trap cleanup EXIT

          export NIX_DISK_IMAGE="$directory/nixos.qcow2"
```

If we remove the `echo` command, then we could vastly simplify the code with:

```suggestion
          # Create a temporary directory for containing the disk image, instead
          # of creating a temporary 'NIX_DISK_IMAGE' file because the
          # 'NIX_DISK_IMAGE' file must not exist as it will otherwise not be
          # written to.
          directory="$(mktemp --directory)"
          trap 'rm --recursive "$directory"' EXIT

          export NIX_DISK_IMAGE="$directory/nixos.qcow2"
```

Reply:
This isn't an activation script, but rather an application which is ran directly by the user, therefore `verboseEcho` won't be defined here.

Running the virtual machine already outputs various text to the console, and our `echo` command mimics the last line as shown below:

```
Disk image do not exist, creating the virtualisation disk image...
Formatting '/tmp/tmp.7m6iQ94vOP', fmt=raw size=1073741824
mke2fs 1.47.0 (5-Feb-2023)
Discarding device blocks: done                            
Creating filesystem with 262144 4k blocks and 65536 inodes
Filesystem UUID: f29b2300-47ef-47d6-81c8-91e38c5358b2
Superblock backups stored on blocks: 
        32768, 98304, 163840, 229376

Allocating group tables: done                            
Writing inode tables: done                            
Creating journal (8192 blocks): done
Writing superblocks and filesystem accounting information: done

Virtualisation disk image created.
Virtualisation disk image removed.
```

Reply:
I've now updated the script to use `/tmp`. The `echo` command is unchanged.

Reply:
*Continued in https://github.com/danth/stylix/pull/320#discussion_r1561846669.*